[Sidecar Containers] Consider init containers in eviction message

toVersus opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

When a Pod is evicted due to resources pressure on the node, the kubelet creates an eviction event and reports it to the user. The event message includes which containers within the Pod exceeded their resource requests, along with their actual usage. If a resource request is not set, it will always be reported as a violation. Similar information is also included in the event annotations.

Currently, the `evictionMessage` function, which generates the message and annotations for the eviction event, does not account for the init containers and restartable init containers (sidecar containers). Since the resources field cannot be specified for ephemeral containers, they will always be blamed for resource overuse when an eviction occurs. That’s why only regular, init and restartable init containers are considered for the eviction message.

This PR improves the eviction message to account for init and restartable init containers. When a Pod with both restartable init containers and regular containers gets evicted, here's what the output looks like:

```yaml
apiVersion: v1
count: 1
eventTime: null
firstTimestamp: "2024-05-19T08:14:14Z"
involvedObject:
  apiVersion: v1
  kind: Pod
  name: alloc-lm9sq
  namespace: default
  resourceVersion: "19979"
  uid: 786af1f6-3b04-4385-b61f-acfca78ddfd6
kind: Event
lastTimestamp: "2024-05-19T08:14:14Z"
message: 'The node was low on resource: memory. Threshold quantity: 1536Mi, available:
  1208704Ki. Container el1 was using 44Ki, request is 0, has larger consumption
  of memory. Container el2 was using 48Ki, request is 0, has larger consumption
  of memory. Container el3 was using 48Ki, request is 0, has larger consumption
  of memory. Container el-sidecar1 was using 3818464Ki, request is 0, has larger
  consumption of memory. Container el-sidecar2 was using 1452Ki, request is 0, has
  larger consumption of memory. Container el-sidecar3 was using 48Ki, request is
  0, has larger consumption of memory. '
metadata:
  annotations:
    offending_containers: el1,el2,el3,el-sidecar1,el-sidecar2,el-sidecar3
    offending_containers_usage: 44Ki,48Ki,48Ki,3818464Ki,1452Ki,48Ki
    starved_resource: memory
  creationTimestamp: "2024-05-19T08:14:14Z"
  name: alloc-lm9sq.17d0d5ba642ef84c
  namespace: default
  resourceVersion: "20072"
  uid: 521c02b7-4935-4def-be2c-53a76c32f4f7
reason: Evicted
reportingComponent: kubelet
reportingInstance: kind-worker
source:
  component: kubelet
  host: kind-worker
type: Warning
```

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #124938

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fixes a bug where restartable and non-restartable init containers were not accounted for in the message and annotations of eviction event.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
[KEP]: https://github.com/kubernetes/enhancements/issues/753
```

toVersus committed:
Consider init containers in eviction message

Signed-off-by: Tsubasa Nagasawa <toversus2357@gmail.com>

toVersus commented:
/cc @SergeyKanzhelev @matthyx @gjkim42

gjkim42 reviewed (commented):
I am not sure it is ok to add restartable init containers to the eviction messages without actually accounting them in eviction logic.

toVersus commented:
Checking the eviction behavior in the local Kubernetes 1.30.0 cluster, I confirmed that eviction occurs even when restartable init containers cause memory pressure on the node.

```sh
cat <<EOF | kind create cluster --config -
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
kubeadmConfigPatches:
- |
  kind: KubeletConfiguration
  evictionHard:
    memory.available: "1.5Gi"
nodes:
- role: control-plane
  image: kindest/node:v1.30.0@sha256:047357ac0cfea04663786a612ba1eaba9702bef25227a794b52890dd8bcd692e
- role: worker
  image: kindest/node:v1.30.0@sha256:047357ac0cfea04663786a612ba1eaba9702bef25227a794b52890dd8bcd692e
EOF

# Create Pod causing eviction by memory pressure
cat <<'EOF' | kubectl create -f -
apiVersion: v1
kind: Pod
metadata: 
  generateName: alloc-
spec:
  terminationGracePeriodSeconds: 1
  restartPolicy: Never
  initContainers:
  - name: el-sidecar1
    image: registry.k8s.io/e2e-test-images/agnhost:2.52
    args: ["stress", "--mem-alloc-size", "100Mi", "--mem-alloc-sleep", "1s", "--mem-total", "4000Mi"]
    env:
    - name: MEMORY_LIMIT
      valueFrom:
        resourceFieldRef:
          resource: limits.memory
    restartPolicy: Always
  - name: el-sidecar2
    image: registry.k8s.io/e2e-test-images/agnhost:2.52
    command: ["sleep", "infinity"]
    restartPolicy: Always
  - name: el-sidecar3
    image: registry.k8s.io/e2e-test-images/agnhost:2.52
    command: ["sleep", "infinity"]
    restartPolicy: Always
  containers:
  - name: el1
    image: registry.k8s.io/e2e-test-images/agnhost:2.52
    command: ["sleep", "infinity"]
  - name: el2
    image: registry.k8s.io/e2e-test-images/agnhost:2.52
    command: ["sleep", "infinity"]
  - name: el3
    image: registry.k8s.io/e2e-test-images/agnhost:2.52
    command: ["sleep", "infinity"]
EOF
```

Details of the evicted Pod:

<details>
<summary><code>kubectl describe pods alloc-5ffbx</code></summary>

```
Name:             alloc-5ffbx
Namespace:        default
Priority:         0
Service Account:  default
Node:             kind-worker/192.168.228.2
Start Time:       Mon, 20 May 2024 21:43:52 +0900
Labels:           <none>
Annotations:      <none>
Status:           Failed
Reason:           Evicted
Message:          The node was low on resource: memory. Threshold quantity: 1536Mi, available: 1248032Ki. Container el3 was using 48Ki, request is 0, has larger consumption of memory. Container el1 was using 44Ki, request is 0, has larger consumption of memory. Container el2 was using 44Ki, request is 0, has larger consumption of memory.
IP:               10.244.1.2
IPs:
  IP:  10.244.1.2
Init Containers:
  el-sidecar1:
    Container ID:
    Image:         registry.k8s.io/e2e-test-images/agnhost:2.52
    Image ID:
    Port:          <none>
    Host Port:     <none>
    Args:
      stress
      --mem-alloc-size
      100Mi
      --mem-alloc-sleep
      1s
      --mem-total
      4000Mi
    State:          Terminated
      Reason:       ContainerStatusUnknown
      Message:      The container could not be located when the pod was terminated
      Exit Code:    137
      Started:      Mon, 01 Jan 0001 00:00:00 +0000
      Finished:     Mon, 01 Jan 0001 00:00:00 +0000
    Last State:     Terminated
      Reason:       ContainerStatusUnknown
      Message:      The container could not be located when the pod was deleted.  The container used to be Running
      Exit Code:    137
      Started:      Mon, 01 Jan 0001 00:00:00 +0000
      Finished:     Mon, 01 Jan 0001 00:00:00 +0000
    Ready:          False
    Restart Count:  1
    Environment:
      MEMORY_LIMIT:  node allocatable (limits.memory)
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-4f6hc (ro)
  el-sidecar2:
    Container ID:  containerd://bf59fcb6ef501a02c929c11b8b702603cb063ec60c60670ade25cbec83881d96
    Image:         registry.k8s.io/e2e-test-images/agnhost:2.52
    Image ID:      registry.k8s.io/e2e-test-images/agnhost@sha256:b173c7d0ffe3d805d49f4dfe48375169b7b8d2e1feb81783efd61eb9d08042e6
    Port:          <none>
    Host Port:     <none>
    Command:
      sleep
      infinity
    State:          Terminated
      Reason:       Error
      Exit Code:    137
      Started:      Mon, 20 May 2024 21:44:05 +0900
      Finished:     Mon, 20 May 2024 21:44:59 +0900
    Ready:          False
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-4f6hc (ro)
  el-sidecar3:
    Container ID:  containerd://1394ee2639ff685ed41da6e1d145432f44051dddcb3ae4fd794c9ae16c4750ba
    Image:         registry.k8s.io/e2e-test-images/agnhost:2.52
    Image ID:      registry.k8s.io/e2e-test-images/agnhost@sha256:b173c7d0ffe3d805d49f4dfe48375169b7b8d2e1feb81783efd61eb9d08042e6
    Port:          <none>
    Host Port:     <none>
    Command:
      sleep
      infinity
    State:          Terminated
      Reason:       Error
      Exit Code:    137
      Started:      Mon, 20 May 2024 21:44:06 +0900
      Finished:     Mon, 20 May 2024 21:44:59 +0900
    Ready:          False
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-4f6hc (ro)
Containers:
  el1:
    Container ID:
    Image:         registry.k8s.io/e2e-test-images/agnhost:2.52
    Image ID:
    Port:          <none>
    Host Port:     <none>
    Command:
      sleep
      infinity
    State:          Terminated
      Reason:       ContainerStatusUnknown
      Message:      The container could not be located when the pod was terminated
      Exit Code:    137
      Started:      Mon, 01 Jan 0001 00:00:00 +0000
      Finished:     Mon, 01 Jan 0001 00:00:00 +0000
    Last State:     Terminated
      Reason:       ContainerStatusUnknown
      Message:      The container could not be located when the pod was deleted.  The container used to be Running
      Exit Code:    137
      Started:      Mon, 01 Jan 0001 00:00:00 +0000
      Finished:     Mon, 01 Jan 0001 00:00:00 +0000
    Ready:          False
    Restart Count:  1
    Environment:    <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-4f6hc (ro)
  el2:
    Container ID:
    Image:         registry.k8s.io/e2e-test-images/agnhost:2.52
    Image ID:
    Port:          <none>
    Host Port:     <none>
    Command:
      sleep
      infinity
    State:          Terminated
      Reason:       ContainerStatusUnknown
      Message:      The container could not be located when the pod was terminated
      Exit Code:    137
      Started:      Mon, 01 Jan 0001 00:00:00 +0000
      Finished:     Mon, 01 Jan 0001 00:00:00 +0000
    Last State:     Terminated
      Reason:       ContainerStatusUnknown
      Message:      The container could not be located when the pod was deleted.  The container used to be Running
      Exit Code:    137
      Started:      Mon, 01 Jan 0001 00:00:00 +0000
      Finished:     Mon, 01 Jan 0001 00:00:00 +0000
    Ready:          False
    Restart Count:  1
    Environment:    <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-4f6hc (ro)
  el3:
    Container ID:
    Image:         registry.k8s.io/e2e-test-images/agnhost:2.52
    Image ID:
    Port:          <none>
    Host Port:     <none>
    Command:
      sleep
      infinity
    State:          Terminated
      Reason:       ContainerStatusUnknown
      Message:      The container could not be located when the pod was terminated
      Exit Code:    137
      Started:      Mon, 01 Jan 0001 00:00:00 +0000
      Finished:     Mon, 01 Jan 0001 00:00:00 +0000
    Last State:     Terminated
      Reason:       ContainerStatusUnknown
      Message:      The container could not be located when the pod was deleted.  The container used to be Running
      Exit Code:    137
      Started:      Mon, 01 Jan 0001 00:00:00 +0000
      Finished:     Mon, 01 Jan 0001 00:00:00 +0000
    Ready:          False
    Restart Count:  1
    Environment:    <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-4f6hc (ro)
Conditions:
  Type                        Status
  DisruptionTarget            True
  PodReadyToStartContainers   False
  Initialized                 False
  Ready                       False
  ContainersReady             False
  PodScheduled                True
Volumes:
  kube-api-access-4f6hc:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   BestEffort
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason     Age   From               Message
  ----     ------     ----  ----               -------
  Normal   Scheduled  82s   default-scheduler  Successfully assigned default/alloc-5ffbx to kind-worker
  Normal   Pulling    82s   kubelet            Pulling image "registry.k8s.io/e2e-test-images/agnhost:2.52"
  Normal   Pulled     69s   kubelet            Successfully pulled image "registry.k8s.io/e2e-test-images/agnhost:2.52" in 12.457s (12.457s including waiting). Image size: 50208832 bytes.
  Normal   Created    69s   kubelet            Created container el-sidecar1
  Normal   Started    69s   kubelet            Started container el-sidecar1
  Normal   Pulled     69s   kubelet            Container image "registry.k8s.io/e2e-test-images/agnhost:2.52" already present on machine
  Normal   Created    69s   kubelet            Created container el-sidecar2
  Normal   Started    69s   kubelet            Started container el-sidecar2
  Normal   Pulled     68s   kubelet            Container image "registry.k8s.io/e2e-test-images/agnhost:2.52" already present on machine
  Normal   Created    68s   kubelet            Created container el-sidecar3
  Normal   Started    68s   kubelet            Started container el-sidecar3
  Normal   Pulled     67s   kubelet            Container image "registry.k8s.io/e2e-test-images/agnhost:2.52" already present on machine
  Normal   Created    67s   kubelet            Created container el1
  Normal   Started    67s   kubelet            Started container el1
  Normal   Pulled     67s   kubelet            Container image "registry.k8s.io/e2e-test-images/agnhost:2.52" already present on machine
  Normal   Created    67s   kubelet            Created container el2
  Normal   Started    67s   kubelet            Started container el2
  Normal   Pulled     67s   kubelet            Container image "registry.k8s.io/e2e-test-images/agnhost:2.52" already present on machine
  Normal   Created    67s   kubelet            Created container el3
  Normal   Started    67s   kubelet            Started container el3
  Warning  Evicted    18s   kubelet            The node was low on resource: memory. Threshold quantity: 1536Mi, available: 1248032Ki. Container el3 was using 48Ki, request is 0, has larger consumption of memory. Container el1 was using 44Ki, request is 0, has larger consumption of memory. Container el2 was using 44Ki, request is 0, has larger consumption of memory.
  Normal   Killing    18s   kubelet            Stopping container el-sidecar1
  Normal   Killing    18s   kubelet            Stopping container el1
  Normal   Killing    18s   kubelet            Stopping container el2
  Normal   Killing    18s   kubelet            Stopping container el3
  Normal   Killing    18s   kubelet            Stopping container el-sidecar3
```

</details>

kannon92 commented:
/ok-to-test

can we add some e2e tests confirming eviction works for sidecars? Memory and storage would be good. A manually test is good but it would be ideal to make sure eviction is tested with sidecars.

it can be a follow up if necessary.

toVersus commented:
I agree that. I initially considered adding E2E tests, but I was reluctant because E2E tests related to Pod eviction are expensive (marked as Slow, Serial, and Disruptive). This is especially true since there is no special care in the eviction logic for sidecar containers. However, if necessary, I am trying to work on adding E2E tests, although I think it might be better to handle it in a separate PR.

bart0sh commented:
/triage accepted
/priority important-longterm

toVersus commented:
/remove-lifecycle stale

This is ready for review.

SergeyKanzhelev commented on pkg/kubelet/eviction/helpers.go:
```diff
@@ -1263,13 +1267,13 @@ func evictionMessage(resourceToReclaim v1.ResourceName, pod *v1.Pod, stats stats
+					exceededContainers = append(exceededContainers, container.Name)
 					containerUsage = append(containerUsage, usage.String())
 				}
 			}
```
why there is no break out of the loop when the name match found?

toVersus replied:
I was wondering about that at first too, but I figured it’s probably because multiple containers might be exceeding their resource requests.

SergeyKanzhelev replied:
Yes, but it will be covered by the outer loop. Inner loop is a lookup of a container by it's name and it should be broken out of when the container with the matching name was found

toVersus replied:
Thanks for the explanation! I get it now. Since both the original implementer and I didn't get it right away, I added a comment, even though it might be a bit redundant.

SergeyKanzhelev commented on pkg/kubelet/eviction/helpers.go:
```diff
@@ -1234,14 +1234,18 @@ func evictionMessage(resourceToReclaim v1.ResourceName, pod *v1.Pod, stats stats
 	}
+	containers := pod.Spec.Containers
+	if len(pod.Spec.InitContainers) != 0 {
+		containers = append(containers, pod.Spec.InitContainers...)
```
oh, I see. there is a comparison with requests and there is no requests for ephemeral contianers. Can you please add this comment here explicitly so next person looking at this code will know why we are not accounting for ephemeral container here.

toVersus replied:
Thanks! As for the ephemeral containers, it was simply my oversight. A you mentioned, specifying the `resources` field results in an error during API server validation, so resource requests can't be set for ephemeral containers. 
Added a comment explaining why ephemeral containers aren’t included here.
https://github.com/kubernetes/kubernetes/blob/f0f7ff989a948389247e628c4c5a43e915f51daa/pkg/apis/core/validation/validation.go#L88

```console
❯ cat <<EOF | kind create cluster --config -
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
- role: control-plane
  image: kindest/node:v1.31.0@sha256:53df588e04085fd41ae12de0c3fe4c72f7013bba32a20e7325357a1ac94ba865
- role: worker
  image: kindest/node:v1.31.0@sha256:53df588e04085fd41ae12de0c3fe4c72f7013bba32a20e7325357a1ac94ba865
EOF
Creating cluster "kind" ...
 ✓ Ensuring node image (kindest/node:v1.31.0) 🖼
 ✓ Preparing nodes 📦 📦
 ✓ Writing configuration 📜
 ✓ Starting control-plane 🕹️
 ✓ Installing CNI 🔌
 ✓ Installing StorageClass 💾
 ✓ Joining worker nodes 🚜
Set kubectl context to "kind-kind"
You can now use your cluster with:

kubectl cluster-info --context kind-kind

Have a nice day! 👋

❯ kubectl version
Client Version: v1.31.1
Kustomize Version: v5.4.2
Server Version: v1.31.0

❯ cat <<'EOF' | kubectl create -f -
apiVersion: v1
kind: Pod
metadata:
  name: alloc
spec:
  terminationGracePeriodSeconds: 1
  restartPolicy: Never
  containers:
  - name: el1
    image: registry.k8s.io/e2e-test-images/agnhost:2.52
    command: ["sleep", "infinity"]
EOF
pod/alloc created

❯ cat <<EOF > custom.json
{
  "resources": {
    "requests": {
      "cpu": "0.2",
      "memory": "256Mi"
    }
  }
}
EOF

❯ kubectl debug alloc -it --custom custom.json --image=registry.k8s.io/e2e-test-images/agnhost:2.52 --target=el1 -- sh
Targeting container "el1". If you don't see processes from this container it may be because the container runtime doesn't support this feature.
Defaulting debug container name to debugger-ch2t6.
The Pod "alloc" is invalid: spec.ephemeralContainers[0].resources: Forbidden: cannot be set for an Ephemeral Container
```

SergeyKanzhelev replied:
cool, thx

toVersus commented:
/retest

toVersus commented:
/retest

SergeyKanzhelev commented:
Thank you. One last nit and one thing to validate and may be change code.

1. nit: it will be cool if you can add unit test with the ephemeral container and will confirm that it will NOT be included into the message.
2. Thing to validate. Is there a chance that the init container that has completed already will be included into the eviction message? Having it there would be rather confusing. If there is no way that stats will be collected and higher than the requests, than it the code is fine as-is. If not - than we might need to add some validation so terminated init containers will not confuse the eviction message

toVersus commented:
Thanks for the suggestion!

> nit: it will be cool if you can add unit test with the ephemeral container and will confirm that it will NOT be included into the message.

Done.

> Thing to validate. Is there a chance that the init container that has completed already will be included into the eviction message?

Looks like the resource usage for terminated containers isn't included in the container stats. I added a debug log in the kubelet, then ran the Pod with a terminated init container and checked the logs, but there’s nothing related to the init containers. 

I updated the test cases for the init containers to make them more realistic:

- A test case for when an eviction occurs while the init container is running
- Revised the other test cases to exclude terminated init containers from the eviction messages

Code change:

```diff
diff --git a/pkg/kubelet/eviction/helpers.go b/pkg/kubelet/eviction/helpers.go
index 4e684a5b902..e6ed91cbabe 100644
--- a/pkg/kubelet/eviction/helpers.go
+++ b/pkg/kubelet/eviction/helpers.go
@@ -1249,8 +1249,10 @@ func evictionMessage(resourceToReclaim v1.ResourceName, pod *v1.Pod, stats stats
                containers = append(containers, pod.Spec.InitContainers...)
        }
        for _, containerStats := range podStats.Containers {
+               klog.InfoS("Eviction manager: check if the container's resource usage exceeds the resource requests", "container", containerStats.Name)
                for _, container := range containers {
                        if container.Name == containerStats.Name {
                                requests := container.Resources.Requests[resourceToReclaim]
                                if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) &&
                                        (resourceToReclaim == v1.ResourceMemory || resourceToReclaim == v1.ResourceCPU) {
```

Pod manifest:

```yaml
cat <<'EOF' | kubectl create -f -
apiVersion: v1
kind: Pod
metadata: 
  generateName: alloc-
spec:
  terminationGracePeriodSeconds: 1
  restartPolicy: Never
  initContainers:
  - name: init1
    image: registry.k8s.io/e2e-test-images/agnhost:2.52
    command: ["sleep", "10"]
  - name: el-sidecar1
    image: registry.k8s.io/e2e-test-images/agnhost:2.52
    args: ["stress", "--mem-alloc-size", "100Mi", "--mem-alloc-sleep", "1s", "--mem-total", "4000Mi"]
    restartPolicy: Always
  - name: el-sidecar2
    image: registry.k8s.io/e2e-test-images/agnhost:2.52
    command: ["sleep", "infinity"]
    restartPolicy: Always
  - name: el-sidecar3
    image: registry.k8s.io/e2e-test-images/agnhost:2.52
    command: ["sleep", "infinity"]
    restartPolicy: Always
  containers:
  - name: el1
    image: registry.k8s.io/e2e-test-images/agnhost:2.52
    command: ["sleep", "infinity"]
  - name: el2
    image: registry.k8s.io/e2e-test-images/agnhost:2.52
    command: ["sleep", "infinity"]
  - name: el3
    image: registry.k8s.io/e2e-test-images/agnhost:2.52
    command: ["sleep", "infinity"]
EOF
```

Kubelet log:

```console
Sep 15 00:07:37 kind-worker kubelet[215]: I0915 00:07:37.162521     215 eviction_manager.go:369] "Eviction manager: attempting to reclaim" resourceName="memory"
Sep 15 00:07:37 kind-worker kubelet[215]: I0915 00:07:37.162993     215 eviction_manager.go:380] "Eviction manager: must evict pod(s) to reclaim" resourceName="memory"
Sep 15 00:07:37 kind-worker kubelet[215]: I0915 00:07:37.163022     215 eviction_manager.go:398] "Eviction manager: pods ranked for eviction" pods=["default/alloc-wpt9g","kube-system/kube-proxy-mhj88","kube-system/kindnet-mvnq8"]
Sep 15 00:07:37 kind-worker kubelet[215]: I0915 00:07:37.168883     215 helpers.go:1252] "Eviction manager: check if the container's resource usage exceeds the resource requests" container="el1"
Sep 15 00:07:37 kind-worker kubelet[215]: I0915 00:07:37.169204     215 helpers.go:1252] "Eviction manager: check if the container's resource usage exceeds the resource requests" container="el3"
Sep 15 00:07:37 kind-worker kubelet[215]: I0915 00:07:37.169221     215 helpers.go:1252] "Eviction manager: check if the container's resource usage exceeds the resource requests" container="el-sidecar1"
Sep 15 00:07:37 kind-worker kubelet[215]: I0915 00:07:37.169241     215 helpers.go:1252] "Eviction manager: check if the container's resource usage exceeds the resource requests" container="el-sidecar2"
Sep 15 00:07:37 kind-worker kubelet[215]: I0915 00:07:37.169257     215 helpers.go:1252] "Eviction manager: check if the container's resource usage exceeds the resource requests" container="el-sidecar3"
Sep 15 00:07:37 kind-worker kubelet[215]: I0915 00:07:37.169270     215 helpers.go:1252] "Eviction manager: check if the container's resource usage exceeds the resource requests" container="el2"
```

toVersus commented:
ping @SergeyKanzhelev

SergeyKanzhelev reviewed (approved):
/approve
/lgtm
